File: sdk/node/scripts/package-runtime.mjs
import {spawnSync} from "node:child_process"
import {chmod, copyFile, mkdir, mkdtemp, readFile, rm, writeFile} from "node:fs/promises"
import {tmpdir} from "node:os"
import {dirname, isAbsolute, join, resolve} from "node:path"
import {fileURLToPath} from "node:url"

const sdkDir = resolve(dirname(fileURLToPath(import.meta.url)), "..")
const repoRoot = resolve(sdkDir, "../..")
const targets = new Set(["linux-x64", "linux-arm64", "darwin-x64", "darwin-arm64", "win32-x64", "win32-arm64"])
const args = parseArgs(process.argv.slice(2))
const target = required(args, "target")
const version = required(args, "version")
const binary = resolve(required(args, "binary"))
const outDir = resolve(required(args, "out-dir"))

if (!targets.has(target)) throw new Error(`unsupported runtime package target ${target}`)
validateVersion(version)

const templateDir = join(sdkDir, "runtime-packages", target)
const template = JSON.parse(await readFile(join(templateDir, "package.json"), "utf8"))
const filename = target.startsWith("win32-") ? "meta-runtime.exe" : "meta-runtime"
const stage = await mkdtemp(join(tmpdir(), "meta-node-runtime-"))

try {
  await mkdir(outDir, {recursive: true})
  await writeFile(join(stage, "package.json"), `${JSON.stringify({...template, version}, null, 2)}\n`)
  await copyFile(binary, join(stage, filename))
  await chmod(join(stage, filename), 0o755)
  await copyFile(join(sdkDir, "runtime-packages", "README.md"), join(stage, "README.md"))
  await copyFile(join(repoRoot, "LICENSE"), join(stage, "LICENSE"))
  const packed = spawnSync("pnpm", ["pack", "--pack-destination", outDir, "--silent"], {cwd: stage, encoding: "utf8"})
  if (packed.status !== 0) throw new Error(`pnpm pack failed: ${(packed.stderr || packed.stdout).trim()}`)
  const output = packed.stdout.trim().split(/\r?\n/).at(-1)
  if (!output) throw new Error("pnpm pack returned no artifact path")
  console.log(isAbsolute(output) ? output : resolve(stage, output))
} finally {
  await rm(stage, {recursive: true, force: true})
}

function parseArgs(values) {
  const result = new Map()
  for (let index = 0; index < values.length; index += 2) {
    const key = values[index]
    const value = values[index + 1]
    if (!key?.startsWith("--") || value === undefined) throw new Error(`invalid argument ${key ?? ""}`)
    result.set(key.slice(2), value)
  }
  return result
}

function required(values, key) {
  const value = values.get(key)
  if (!value) throw new Error(`--${key} is required`)
  return value
}

function validateVersion(version) {
  if (!/^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?(?:\+[0-9A-Za-z.-]+)?$/.test(version)) throw new Error(`invalid package version ${version}`)
}

File: sdk/python/scripts/package-runtime.py
from __future__ import annotations

import argparse
import base64
import csv
import hashlib
import io
import re
import stat
import zipfile
from pathlib import Path

TARGETS = {
    "linux-x64-manylinux": "manylinux_2_17_x86_64",
    "linux-x64-musllinux": "musllinux_1_2_x86_64",
    "linux-arm64-manylinux": "manylinux_2_17_aarch64",
    "linux-arm64-musllinux": "musllinux_1_2_aarch64",
    "darwin-x64": "macosx_11_0_x86_64",
    "darwin-arm64": "macosx_11_0_arm64",
    "win32-x64": "win_amd64",
    "win32-arm64": "win_arm64",
}
DIST = "mewisme_meta_runtime"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--target", required=True, choices=sorted(TARGETS))
    parser.add_argument("--version", required=True)
    parser.add_argument("--binary", required=True, type=Path)
    parser.add_argument("--out-dir", required=True, type=Path)
    args = parser.parse_args()
    validate_version(args.version)
    binary = args.binary.resolve()
    if not binary.is_file():
        raise SystemExit(f"runtime binary does not exist: {binary}")
    out = build_wheel(args.target, args.version, binary, args.out_dir.resolve())
    print(out)


def build_wheel(target: str, version: str, binary: Path, out_dir: Path) -> Path:
    platform_tag = TARGETS[target]
    filename = "meta-runtime.exe" if target.startswith("win32-") else "meta-runtime"
    wheel = out_dir / f"{DIST}-{version}-py3-none-{platform_tag}.whl"
    dist_info = f"{DIST}-{version}.dist-info"
    repo_root = Path(__file__).resolve().parents[3]
    locator = (Path(__file__).resolve().parents[1] / "runtime" / "src" / DIST / "__init__.py").read_bytes()
    files = {
        f"{DIST}/__init__.py": (locator, 0o644),
        f"{DIST}/{filename}": (binary.read_bytes(), 0o755),
        f"{dist_info}/METADATA": (metadata(version).encode(), 0o644),
        f"{dist_info}/WHEEL": (wheel_metadata(platform_tag).encode(), 0o644),
        f"{dist_info}/licenses/LICENSE": ((repo_root / "LICENSE").read_bytes(), 0o644),
    }
    out_dir.mkdir(parents=True, exist_ok=True)
    records = []
    with zipfile.ZipFile(wheel, "w", compression=zipfile.ZIP_DEFLATED, compresslevel=9) as archive:
        for name, (data, mode) in files.items():
            write_member(archive, name, data, mode)
            records.append((name, wheel_hash(data), str(len(data))))
        record_name = f"{dist_info}/RECORD"
        record = record_csv([*records, (record_name, "", "")]).encode()
        write_member(archive, record_name, record, 0o644)
    return wheel


def metadata(version: str) -> str:
    return (
        "Metadata-Version: 2.4\n"
        "Name: mewisme-meta-runtime\n"
        f"Version: {version}\n"
        "Summary: Platform runtime for mewisme-meta\n"
        "Requires-Python: >=3.10\n"
        "License-Expression: AGPL-3.0-only\n"
        "Project-URL: Repository, https://github.com/mewisme/meta.go\n"
        "\n"
    )


def wheel_metadata(platform_tag: str) -> str:
    return "Wheel-Version: 1.0\nGenerator: meta.go\nRoot-Is-Purelib: false\nTag: py3-none-" + platform_tag + "\n"


def write_member(archive: zipfile.ZipFile, name: str, data: bytes, mode: int) -> None:
    info = zipfile.ZipInfo(name, (2020, 1, 1, 0, 0, 0))
    info.create_system = 3
    info.external_attr = (stat.S_IFREG | mode) << 16
    info.compress_type = zipfile.ZIP_DEFLATED
    archive.writestr(info, data)


def wheel_hash(data: bytes) -> str:
    digest = base64.urlsafe_b64encode(hashlib.sha256(data).digest()).rstrip(b"=").decode()
    return f"sha256={digest}"


def record_csv(rows: list[tuple[str, str, str]]) -> str:
    output = io.StringIO(newline="")
    csv.writer(output, lineterminator="\n").writerows(rows)
    return output.getvalue()


def validate_version(version: str) -> None:
    if re.fullmatch(r"\d+(?:\.\d+){2,}(?:(?:a|b|rc)\d+)?", version) is None:
        raise SystemExit(f"invalid package version: {version}")


if __name__ == "__main__":
    main()
